### ISSUE TITLE
udisks_client_get_block_for_drive() always returns the wrong block of eMMC

### ISSUE THREAD

**Issue by starnight** [storaged-project/udisks#879]
We have an ASUS E406MA equipped with an eMMC.  The eMMC has blocks: `/dev/mmcblk0`, `/dev/mmcblk0boot0`, `/dev/mmcblk0boot1`, and `/dev/mmcblk0rpmb`.
We notice [udisks_client_get_block_for_drive()](https://github.com/storaged-project/udisks/blob/85584a1be5edad87a3b7e885bac38508f8917b69/udisks/udisksclient.c#L894) always returns the wrong block `/dev/mmcblk0boot1` of eMMC which leads other programs get wrong behavior.

I prepare the small reproducer and corresponding Makefile: [test-udisks.tar.gz](https://github.com/storaged-project/udisks/files/6349176/test-udisks.tar.gz)
```
$ make
gcc main.c `pkg-config --cflags --libs udisks2 glib-2.0` -Wall -g -o get_block
$ ./get_block 
/dev/sda 
/dev/mmcblk0boot1
```
Here shows `/dev/mmcblk0boot1`, not the expected `/dev/mmcblk0`.

**Comment by starnight** [comment]
Let's go through with gdb:
* [get_top_level_blocks_for_drive()](https://github.com/storaged-project/udisks/blob/85584a1be5edad87a3b7e885bac38508f8917b69/udisks/udisksclient.c#L820) found and added 3 blocks `/dev/mmcblk0boot1`, `/dev/mmcblk0` and `/dev/mmcblk0boot0` of the eMMC drive `/org/freedesktop/UDisks2/drives/DA4064_0x65530d06` into the block list.
```
Thread 1 "get_block" hit Breakpoint 3, get_top_level_blocks_for_drive (drive_object_path=0x5555555bf400 "/org/freedesktop/UDisks2/drives/DA4064_0x65530d06", client=<optimized out>) at udisksclient.c:846
846	          ret = g_list_append (ret, g_object_ref (object));
(gdb) print udisks_block_get_device(block)
$2 = (const gchar *) 0x7fffe805b0e0 "/dev/mmcblk0boot1"
(gdb) continue
Continuing.

Thread 1 "get_block" hit Breakpoint 3, get_top_level_blocks_for_drive (drive_object_path=0x5555555bf400 "/org/freedesktop/UDisks2/drives/DA4064_0x65530d06", client=<optimized out>) at udisksclient.c:846
846	          ret = g_list_append (ret, g_object_ref (object));
(gdb) print udisks_block_get_device(block)
$3 = (const gchar *) 0x7fffe80424f0 "/dev/mmcblk0"
(gdb) continue
Continuing.

Thread 1 "get_block" hit Breakpoint 3, get_top_level_blocks_for_drive (drive_object_path=0x5555555bf400 "/org/freedesktop/UDisks2/drives/DA4064_0x65530d06", client=<optimized out>) at udisksclient.c:846
846	          ret = g_list_append (ret, g_object_ref (object));
(gdb) print udisks_block_get_device(block)
$4 = (const gchar *) 0x7fffe803ee10 "/dev/mmcblk0boot0"
(gdb) continue
Continuing.

Thread 1 "get_block" hit Breakpoint 4, get_top_level_blocks_for_drive (drive_object_path=0x5555555bf400 "/org/freedesktop/UDisks2/drives/DA4064_0x65530d06", client=<optimized out>) at udisksclient.c:850
850	  g_list_free_full (object_proxies, g_object_unref);
(gdb) finish
```
* But, [udisks_client_get_block_for_drive()](https://github.com/storaged-project/udisks/blob/85584a1be5edad87a3b7e885bac38508f8917b69/udisks/udisksclient.c#L894) only picks `/dev/mmcblk0boot1` to return, instead of `/dev/mmcblk0`.
```
Run till exit from #0  get_top_level_blocks_for_drive (drive_object_path=0x5555555bf400 "/org/freedesktop/UDisks2/drives/DA4064_0x65530d06", client=<optimized out>) at udisksclient.c:850
udisks_client_get_block_for_drive (client=<optimized out>, drive=<optimized out>, get_physical=<optimized out>) at udisksclient.c:892
892	  for (l = blocks; l != NULL; l = l->next)
(gdb) list
887	  object = g_dbus_interface_get_object (G_DBUS_INTERFACE (drive));
888	  if (object == NULL)
889	    goto out;
890	
891	  blocks = get_top_level_blocks_for_drive (client, g_dbus_object_get_object_path (object));
892	  for (l = blocks; l != NULL; l = l->next)
893	    {
894	      UDisksBlock *block = udisks_object_peek_block (UDISKS_OBJECT (l->data));
895	      if (block != NULL)
896	        {
(gdb) list
897	          /* TODO: actually look at @get_physical */
898	          ret = g_object_ref (block);
899	          goto out;
900	        }
901	    }
902	
903	 out:
904	  g_list_free_full (blocks, g_object_unref);
905	  return ret;
906	}
(gdb) break udisksclient.c:895
Breakpoint 5 at 0x7ffff7f6dadd: file udisksclient.c, line 895.
(gdb) continue
Continuing.

Thread 1 "get_block" hit Breakpoint 5, udisks_client_get_block_for_drive (client=<optimized out>, drive=<optimized out>, get_physical=<optimized out>) at udisksclient.c:895
895	      if (block != NULL)
(gdb) print udisks_block_get_device(block)
$5 = (const gchar *) 0x7fffe805b0e0 "/dev/mmcblk0boot1"
(gdb) next
898	          ret = g_object_ref (block);
(gdb) next
904	  g_list_free_full (blocks, g_object_unref);
(gdb) finish
```
Here is the full gdb log: [gdb.log](https://github.com/storaged-project/udisks/files/6349223/gdb.log)

**Comment by starnight** [comment]
Check the eMMC with `fdisk`.  `/dev/mmcblk0boot0` and `/dev/mmcblk0boot1` are not the partitions belonging to `/dev/mmcblk0`, nor belonging to each other.
```
$ sudo fdisk -l /dev/mmcblk0
Disk /dev/mmcblk0: 58.24 GiB, 62537072640 bytes, 122142720 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: gpt
Disk identifier: F7478553-E4EF-1B4F-A6CA-CEA97818C474

Device          Start       End   Sectors  Size Type
/dev/mmcblk0p1   2048    129023    126976   62M EFI System
/dev/mmcblk0p2 129024    131071      2048    1M BIOS boot
/dev/mmcblk0p3 131072 122142686 122011615 58.2G Linux root (x86-64)
$ sudo fdisk -l /dev/mmcblk0boot0 
Disk /dev/mmcblk0boot0: 4 MiB, 4194304 bytes, 8192 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
$ sudo fdisk -l /dev/mmcblk0boot1
Disk /dev/mmcblk0boot1: 4 MiB, 4194304 bytes, 8192 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
```
Maybe, `/dev/mmcblk0`, `/dev/mmcblk0boot0` and `/dev/mmcblk0boot1` should belong to three different **drive**s by each.

**Comment by wjt** [comment]
This is a side-effect of the fix in #834 for the issue I reported in #619. The goal there was to not have separate drives for the boot partitions. The motivation was that [eos-installer](https://github.com/endlessm/eos-installer/) presents the user with a list of Drives to install the OS to, and before #834 an additional 4 MiB drive was offered in that list. This is also the reason the current behaviour is problematic for us: eos-installer uses `udisks_client_get_block_for_drive()` followed by `OpenForRestore()` to get a file handle to install the OS to, and due to this bug it tries (and fails) to install the OS to a hardware partition.

In my opinion there are two problems here:

1. get_block_for_drive is not deterministic: it returns whatever happens to be first in the list of objects returned by `udisks_client_get_block_for_drive()`
2. as a result it can return the wrong block in cases where there is more than 1.

**Comment by wjt** [comment]
#881 has a possible fix to this function.
